Framework/Algorithms/src/Bin2DPowderDiffraction.cpp

```diff
@@ -0,0 +1,385 @@
+      outputWS->setBinEdges(idx, binEdges);
+      idx++;
+    }
+
+  } else {
+    static_cast<void>(createAxisFromRebinParams(getProperty("Axis1Binning"),
+                                                newXBins.mutableRawData()));
+    HistogramData::BinEdges binEdges(newXBins);
+    newYSize = createAxisFromRebinParams(getProperty("Axis2Binning"), newY);
+    newXSize = binEdges.size();
+    outputWS = WorkspaceFactory::Instance().create(m_inputWS, newYSize - 1,
+                                                   newXSize, newXSize - 1);
+    for (size_t idx = 0; idx < newYSize - 1; idx++)
+      outputWS->setBinEdges(idx, binEdges);
+    NumericAxis *const abscissa = new BinEdgeAxis(newXBins.mutableRawData());
+    outputWS->replaceAxis(0, abscissa);
+  }
+
+  outputWS->getAxis(0)->unit() = UnitFactory::Instance().create("dSpacing");
+
+  NumericAxis *const verticalAxis = new BinEdgeAxis(newY);
+  // Meta data
+  verticalAxis->unit() =
+      UnitFactory::Instance().create("dSpacingPerpendicular");
+  verticalAxis->title() = "d_p";
+  outputWS->replaceAxis(1, verticalAxis);
+
+  Progress prog(this, 0.0, 1.0, m_numberOfSpectra);
+  int64_t numSpectra = static_cast<int64_t>(m_numberOfSpectra);
+  std::vector<std::vector<double>> newYValues(
+      newYSize - 1, std::vector<double>(newXSize - 1, 0.0));
+  std::vector<std::vector<double>> newEValues(
+      newYSize - 1, std::vector<double>(newXSize - 1, 0.0));
+
+  // fill the workspace with data
+  g_log.debug() << "newYSize = " << newYSize << std::endl;
+  g_log.debug() << "newXSize = " << newXSize << std::endl;
+  std::vector<double> dp_vec(verticalAxis->getValues());
+
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputWS, *outputWS))
+  for (int64_t snum = 0; snum < numSpectra; ++snum) {
+    PARALLEL_START_INTERUPT_REGION
+    if (!spectrumInfo.isMasked(snum)) {
+      double theta = 0.5 * spectrumInfo.twoTheta(snum);
+      if (theta == 0) {
+        throw std::runtime_error("Spectrum " + std::to_string(snum) +
+                                 " has theta=0. Cannot calculate d-Spacing!");
+      }
+      if (cos(theta) <= 0) {
+        throw std::runtime_error(
+            "Spectrum " + std::to_string(snum) +
+            " has cos(theta) <= 0. Cannot calculate d-SpacingPerpendicular!");
+      }
+      EventList &evList = m_inputWS->getSpectrum(snum);
+
+      // Switch to weighted if needed.
+      if (evList.getEventType() == TOF)
+        evList.switchTo(WEIGHTED);
+
+      std::vector<WeightedEvent> events = evList.getWeightedEvents();
+
+      for (const auto &ev : events) {
+        double d, dp;
+        convertToDSpacing(ev.tof(), theta, &d, &dp);
+        std::vector<double>::iterator upy =
+            std::lower_bound(dp_vec.begin(), dp_vec.end(), dp);
+        int64_t h_index = std::distance(dp_vec.begin(), upy) - 1;
+        if ((h_index < static_cast<int>(newYSize) - 1) && h_index > -1) {
+          if (h_index == static_cast<int>(newYSize) - 1)
```

Comment:
`static_cast<int64_t>` to minimize reducing size of `newYSize` which is `size_t`